File: backends/go-gin/internal/infrastructure/cache/redis_gate_repository.go
package cache

import (
	"authorization-go/internal/config"
	"context"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
)

// RedisGateRepository は domain/gate.CacheRepository の Redis 実装です。
type RedisGateRepository struct {
	rdb *redis.Client
	cfg *config.Config
}

// NewRedisGateRepository は RedisGateRepository を生成します。
//
// rdb: Redis クライアント
// cfg: アプリケーション設定
func NewRedisGateRepository(rdb *redis.Client, cfg *config.Config) *RedisGateRepository {
	return &RedisGateRepository{rdb: rdb, cfg: cfg}
}

// GetJwt はキャッシュからゲート JWT を取得します。キャッシュミスの場合は空文字を返します。
func (r *RedisGateRepository) GetJwt(identifier, memberID string) (string, error) {
	ctx := context.Background()
	val, err := r.rdb.Get(ctx, r.key(identifier, memberID)).Result()
	if err == redis.Nil {
		return "", nil
	}
	return val, err
}

// PutJwt はゲート JWT をキャッシュに保存します。
func (r *RedisGateRepository) PutJwt(identifier, memberID, token string, ttl int) error {
	ctx := context.Background()
	return r.rdb.Set(ctx, r.key(identifier, memberID), token, time.Duration(ttl)*time.Second).Err()
}

func (r *RedisGateRepository) key(identifier, memberID string) string {
	prefix := r.cfg.App.CachePrefix
	if prefix == "" {
		return fmt.Sprintf("gate.jwt:%s:%s", identifier, memberID)
	}
	return fmt.Sprintf("%s:gate.jwt:%s:%s", prefix, identifier, memberID)
}

File: backends/ruby-rails/app/infrastructure/cache/redis_gate_repository.rb
# frozen_string_literal: true
#
# Redis を用いたゲートキャッシュリポジトリの実装モジュール。
#
# @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>

module Infrastructure
  module Cache
    # Redis を用いたゲートキャッシュリポジトリの実装クラスです。
    # @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
    class RedisGateRepository
      # @param cfg [AppConfig] アプリケーション設定
      def initialize(cfg)
        @redis  = RedisClient.new_client(cfg)
        @prefix = cfg.app.cache_prefix
      end

      def get_jwt(identifier, member_id)
        @redis.get(cache_key(identifier, member_id))
      end

      def put_jwt(identifier, member_id, token, ttl)
        @redis.set(cache_key(identifier, member_id), token, ex: ttl)
        nil
      end

      private

      def cache_key(identifier, member_id)
        "#{@prefix}:gate.jwt:#{identifier}:#{member_id}"
      end
    end
  end
end
